File: algorithms/stack/src/Prefix2Postfix.java
import java.util.Stack;

/**
 * @author Santhosh Kumar Tekuri
 */
class Prefix2Postfix {
    static String toPostfix(String prefix) {
        Stack<String> stack = new Stack<>();
        for (int i = prefix.length() - 1; i >= 0; i--) {
            char ch = prefix.charAt(i);
            if (isOperator(ch)) {
                String op1 = stack.pop();
                String op2 = stack.pop();
                stack.push(op1 + op2 + String.valueOf(ch));
            } else
                stack.push(String.valueOf(ch));
        }
        return stack.pop();
    }

    static boolean isOperator(char ch) {
        switch (ch) {
            case '+':
            case '-':
            case '*':
            case '/':
                return true;
            default:
                return false;
        }
    }

    public static void main(String[] args) {
        System.out.println(toPostfix("+-435"));
        System.out.println(toPostfix("-*+ABC*-DE+FG"));
    }
}

File: algorithms/stack/src/Prefix2Postfix.go
package main

import (
	"fmt"
)

func toPostfix(prefix string) string {
	var stack stringStack
	for i := len(prefix) - 1; i >= 0; i-- {
		switch ch := prefix[i]; ch {
		case '+', '-', '*', '/':
			op1 := stack.pop()
			op2 := stack.pop()
			stack.push(op1 + op2 + string(ch))
		default:
			stack.push(string(ch))
		}
	}
	return stack.peek()
}

func main() {
	fmt.Println(toPostfix("+-435"))
	fmt.Println(toPostfix("-*+ABC*-DE+FG"))
}

//----------------------------------------------------------

type stringStack struct {
	arr []string
}

func (s *stringStack) push(v string) {
	s.arr = append(s.arr, v)
}

func (s *stringStack) pop() string {
	v := s.arr[len(s.arr)-1]
	s.arr = s.arr[:len(s.arr)-1]
	return v
}

func (s *stringStack) peek() string {
	return s.arr[len(s.arr)-1]
}
